### PR TITLE
Quality Check : Add SpEL expression

### PR DESCRIPTION
**Pull request by fviale**
- new syntax PA:SPEL to use Spring Expression Language in models
   - expression has access to job and tasks variables to do multiple variables checks
   - expression can eventually modify other variables
    - method validate can now return in JobValidationData a map of updated variables, if the validation made some modification to them.

### CODE REVIEW

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 2 issues

1. MINOR [ParserValidator.java#L55](https://github.com/ow2-proactive/scheduling/blob/36b6dccf96dbb8c14fdc5139f72ab4fd0449d9c1/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/factory/ParserValidator.java#L55): Refactor this method to throw at most one checked exception instead of: org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ConversionException, org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ValidationException, org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ModelSyntaxException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)
1. MINOR [CRONValidator.java#L37](https://github.com/ow2-proactive/scheduling/blob/36b6dccf96dbb8c14fdc5139f72ab4fd0449d9c1/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/validator/CRONValidator.java#L37): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1186)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 1 issue

1. MINOR [ParserValidator.java#L55](https://github.com/ow2-proactive/scheduling/blob/36b6dccf96dbb8c14fdc5139f72ab4fd0449d9c1/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/factory/ParserValidator.java#L55): Refactor this method to throw at most one checked exception instead of: org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ConversionException, org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ValidationException, org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ModelSyntaxException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/ValidationUtil.java [hunk 1]
```diff
@@ -106,6 +110,19 @@ private static JobValidationData validateJob(String jobFilePath) {
...
+        HashMap<String, String> updatedVariables = new HashMap<>();
+        for (JobVariable jobVariable : job.getVariables().values()) {
+            updatedVariables.put(jobVariable.getName(), jobVariable.getValue());
+        }
+        for (Task task : job.getTasks()) {
+            for (TaskVariable taskVariable : task.getVariables().values()) {
+                updatedVariables.put(task.getName() + ":" + taskVariable.getName(), taskVariable.getValue());
+            }
+        }
+        data.setUpdatedVariables(updatedVariables);
```
Comment (marcocast) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/ValidationUtil.java]: is it possible that data.updatedVariables already contains some value before this point? Basically is there the risk to loose/reset information here?

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/JobValidatorService.java [hunk 2]
```diff
@@ -45,18 +45,20 @@
      * Example of use: xml schema validation.
      *
      * @param jobFile xml job file to validate
+     * @return if the validator eventually made some modifications to the job, return a new version
      * @throws JobValidationException if the job is not valid
      */
-    void validateJob(File jobFile) throws JobValidationException;
+    File validateJob(File jobFile) throws JobValidationException;
```
Comment (marcocast) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/JobValidatorService.java]: a validation method should not return any object

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/JobValidatorService.java [hunk 3]
```diff
@@ -45,18 +45,20 @@
...
      * Validate a job object after the job has been parsed by the scheduler.
      *
      * Example of use: variable model validation.
      *
      * @param job job object to validate
+     * @return if the validator eventually made some modifications to the job, return a new version
      * @throws JobValidationException if the job is not valid
      */
-    void validateJob(TaskFlowJob job) throws JobValidationException;
+    TaskFlowJob validateJob(TaskFlowJob job) throws JobValidationException;
```
Comment (marcocast) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/JobValidatorService.java]: a validation method should not return any object

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/ModelValidatorContext.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+        for (Task task : job.getTasks()) {
+            Map<String, Serializable> taskVariablesValues = new LinkedHashMap<>();
+            for (TaskVariable taskVariable : task.getVariables().values()) {
+                taskVariablesValues.put(taskVariable.getName(), taskVariable.getValue());
+            }
+            tasks.put(task.getName(), new SPELTaskVariables(taskVariablesValues));
+        }
+
+        spELJobAndTaskVariables = new SpELJobAndTaskVariables(jobVariablesValues, tasks);
```
Comment (marcocast) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/ModelValidatorContext.java]: SpELJobAndTaskVariables  with p lowercase?

**Comment by paraita** [comment]
Also, there's no way a user can inject arbitrary code with SPEL ?

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/ValidationUtil.java (hunk 1 above)
Comment (fviale) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/ValidationUtil.java]: It is supposed to fill updated variables according to the job in parameter, so even if there were already a map of variables in the validation data, it must be overwritten

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/JobValidatorService.java (hunk 2 above)
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/JobValidatorService.java]: Here is an example of validation methods which return objects https://commons.apache.org/proper/commons-lang/javadocs/api-3.1/org/apache/commons/lang3/Validate.html

File: scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/ModelValidatorContext.java (hunk 4 above)
Comment (fviale) [inline scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/ModelValidatorContext.java]: I tried to match the naming syntax https://docs.spring.io/spring/docs/current/spring-framework-reference/html/expressions.html

Unfortunately, I did not manage to respect it for some file names due to windows file name case insensitivity.

**Comment by fviale** [comment]
@paraita, I will leave your last comment to a later work

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 1 issue

1. MINOR [ParserValidator.java#L55](https://github.com/ow2-proactive/scheduling/blob/42d68061fe260e27e75ce0401135d4ae00652f8b/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/job/factories/spi/model/factory/ParserValidator.java#L55): Refactor this method to throw at most one checked exception instead of: org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ConversionException, org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ValidationException, org.ow2.proactive.scheduler.common.job.factories.spi.model.exceptions.ModelSyntaxException [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1160)

**Pull request merged**